Amethyst/Layout/ThreeColumnLayout.swift

```diff
@@ -0,0 +1,270 @@
+            let column: Column = paneArrangement.column(for: pane)
+
+            let (leftPaneWidth, middlePaneWidth, _): (CGFloat, CGFloat, CGFloat) = paneArrangement.widthsLeftToRight()
+
+            let xorigin: CGFloat = {
+                switch column {
+                case .left: return screenFrame.origin.x
+                case .middle: return screenFrame.origin.x + leftPaneWidth
+                case .right: return screenFrame.origin.x + leftPaneWidth + middlePaneWidth
+                }
+            }()
+
+            let scaleFactor: CGFloat = (screenFrame.width / windowWidth)
+
+            windowFrame.origin.x = xorigin
+            windowFrame.origin.y = screenFrame.origin.y + (windowHeight * CGFloat(paneIndex))
+            windowFrame.size.width = windowWidth
+            windowFrame.size.height = windowHeight
+
+            let isTertiaryMain = (paneArrangement.tertiaryPaneCount > 0 ? pane == Pane.main : pane != Pane.main)
+
+            let resizeRules = ResizeRules(isMain: isTertiaryMain, unconstrainedDimension: .horizontal, scaleFactor: scaleFactor)
+            let frameAssignment = FrameAssignment(frame: windowFrame, window: window, focused: window.isEqual(to: focusedWindow), screenFrame: screenFrame, resizeRules: resizeRules)
+
+            assignments.append(frameAssignment)
+
+            return assignments
+        }
+    }
+
+    override func main() {
+        guard !isCancelled else {
+            return
+        }
+
+        layout.performFrameAssignments(frameAssignments())
+    }
+}
+
+// not an actual Layout, just a base class for the three actual Layouts below
+class ThreeColumnLayout: WindowActivityCache {
+    let windowActivityCache: WindowActivityCache
+
+    init(windowActivityCache: WindowActivityCache) {
+        self.windowActivityCache = windowActivityCache
+    }
+
+    func windowIsActive(_ window: SIWindow) -> Bool {
+        return windowActivityCache.windowIsActive(window)
+    }
+
+    fileprivate var mainPaneCount: Int = 1
+    fileprivate(set) var mainPaneRatio: CGFloat = 0.5
+}
+
+final class ThreeColumnLeftLayout: ThreeColumnLayout, Layout {
+    static var layoutName: String { return "3Column Left" }
+    static var layoutKey: String { return "3column-left" }
+
+    func reflow(_ windows: [SIWindow], on screen: NSScreen) -> ReflowOperation {
+        return ThreeColumnReflowOperation(screen: screen, windows: windows, layout: self, frameAssigner: self, mainIs: Column.left)
+    }
+
+    func assignedFrame(_ window: SIWindow, of windows: [SIWindow], on screen: NSScreen) -> FrameAssignment? {
+        return ThreeColumnReflowOperation(screen: screen, windows: windows, layout: self, frameAssigner: self, mainIs: Column.left).frameAssignments().first { $0.window == window }
+    }
+}
+
+final class ThreeColumnMiddleLayout: ThreeColumnLayout, Layout {
+    static var layoutName: String { return "3Column Middle" }
+    static var layoutKey: String { return "3column-middle" }
```

Comment:
Would that make the layout show as "middle-wide" in the Preferences dialog though?  I think it would be confusing if the name in the Preferences dialog didn't match the name in the HUD.  (Unless I'm wrong about that.)

An alternate solution would be to show the HUD names in the Preferences dialog, instead of the keys.  Then the key would be purely internal.

Reply:
That can be a separate change, though, if you'd like to keep this change more self-contained.

Reply:
Sounds good.  I un-renamed the key to `middle-wide` while leaving the name as `3Column Middle`, and I'll work on a separate PR to have Preferences start using names.